File: typescript/src/vcs.ts
/**
 * Spatial version control system (VCS) operations.
 *
 * Provides helpers for managed repository creation, commit history, and diffs.
 * @module vcs
 */

import type { RoteiroClient } from './client';
import type { Commit, DiffResult, Repo } from './types';

export interface CreateRepoOptions {
  projectId?: number;
  datasetName?: string;
}

export interface ListReposOptions {
  projectId?: number;
}

function resolveProjectId(
  client: RoteiroClient,
  projectId?: number,
): number | undefined {
  return projectId ?? client.getProjectId();
}

/**
 * Create a managed VCS repository.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param name - Display name for the repository.
 * @param options - Optional project/dataset linkage.
 * @returns The created managed repository record.
 */
export async function createRepo(
  client: RoteiroClient,
  name: string,
  options: CreateRepoOptions = {},
): Promise<Repo> {
  const body: {
    name: string;
    project_id?: number;
    dataset_name?: string;
  } = { name };
  const projectId = resolveProjectId(client, options.projectId);
  if (projectId !== undefined) {
    body.project_id = projectId;
  }
  if (options.datasetName) {
    body.dataset_name = options.datasetName;
  }
  return client.post<Repo>('/api/vcs/repos', body);
}

/**
 * Backward-compatible alias for {@link createRepo}.
 *
 * Older SDK versions used `initRepo` for a filesystem-path-based flow that is
 * no longer exposed by the server. The current API creates managed
 * repositories by name instead.
 */
export async function initRepo(
  client: RoteiroClient,
  name: string,
  options: CreateRepoOptions = {},
): Promise<Repo> {
  return createRepo(client, name, options);
}

/**
 * List managed VCS repositories visible to the current tenant.
 */
export async function listRepos(
  client: RoteiroClient,
  options: ListReposOptions = {},
): Promise<Repo[]> {
  const projectId = resolveProjectId(client, options.projectId);
  const params = new URLSearchParams();
  if (projectId !== undefined) {
    params.set('project_id', String(projectId));
  }
  const query = params.toString();
  return client.request<Repo[]>(`/api/vcs/repos${query ? `?${query}` : ''}`);
}

/**
 * Fetch a managed VCS repository by ID.
 */
export async function getRepo(
  client: RoteiroClient,
  repoId: string,
): Promise<Repo> {
  return client.request<Repo>(`/api/vcs/repos/${encodeURIComponent(repoId)}`);
}

/**
 * Delete a managed VCS repository by ID.
 */
export async function deleteRepo(
  client: RoteiroClient,
  repoId: string,
): Promise<void> {
  await client.del(`/api/vcs/repos/${encodeURIComponent(repoId)}`);
}

/**
 * Create a new commit in a VCS repository.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param repoId - Managed repository ID.
 * @param inputPath - Path to the geospatial dataset to commit.
 * @param message - Human-readable commit message.
 * @returns The created Commit object.
 */
export async function commit(
  client: RoteiroClient,
  repoId: string,
  inputPath: string,
  message: string,
): Promise<Commit> {
  return client.post<Commit>('/api/vcs/commit', {
    repo_id: repoId,
    input: inputPath,
    message,
  });
}

/**
 * Get the commit history for a repository.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param repoId - Managed repository ID.
 * @returns Commits ordered from most recent to oldest.
 */
export async function log(client: RoteiroClient, repoId: string): Promise<Commit[]> {
  return client.request<Commit[]>(
    `/api/vcs/log?repo_id=${encodeURIComponent(repoId)}`,
  );
}

/**
 * Get the commit history for a dataset's managed repository.
 */
export async function logForDataset(
  client: RoteiroClient,
  datasetName: string,
): Promise<Commit[]> {
  return client.request<Commit[]>(
    `/api/vcs/log/dataset/${encodeURIComponent(datasetName)}`,
  );
}

/**
 * Compute the diff between two commits in a managed repository.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param repoId - Managed repository ID.
 * @param commitA - Commit ID for the base (from) commit.
 * @param commitB - Commit ID for the target (to) commit.
 * @returns A DiffResult with added, removed, and modified features.
 */
export async function diff(
  client: RoteiroClient,
  repoId: string,
  commitA: string,
  commitB: string,
): Promise<DiffResult> {
  const params = new URLSearchParams({
    repo_id: repoId,
    from: commitA,
    to: commitB,
  });
  return client.request<DiffResult>(`/api/vcs/diff?${params.toString()}`);
}

/**
 * Check out a specific commit (retrieve a historical snapshot).
 *
 * This is a convenience function that retrieves the diff from the initial
 * commit to the target commit. Full snapshot restoration may require
 * reading the commit blob directly.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param repoId - Managed repository ID.
 * @param commitId - The commit ID to check out.
 */
export async function checkout(
  client: RoteiroClient,
  repoId: string,
  commitId: string,
): Promise<void> {
  const commits = await log(client, repoId);
  if (commits.length === 0) {
    throw new Error('No commits in repository');
  }
  const oldest = commits[commits.length - 1];
  if (oldest.id === commitId) {
    return;
  }
  await diff(client, repoId, oldest.id, commitId);
}

File: python/roteiro/vcs.py
"""Spatial version control system (VCS) operations.

Provides helpers for managed repository creation, commit history, and diffs.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, List, Optional

from .client import _encode_path_value, _with_query
from .models import Commit, DiffResult, Repo

if TYPE_CHECKING:
    from .client import RoteiroClient


def create_repo(
    client: RoteiroClient,
    name: str,
    project_id: Optional[int] = None,
    dataset_name: Optional[str] = None,
) -> Repo:
    """Create a managed VCS repository.

    Args:
        client: An initialised RoteiroClient instance.
        name: Repository display name.
        project_id: Optional project association. Defaults to the client's
            configured project scope when available.
        dataset_name: Optional dataset linkage for the repository.

    Returns:
        The created repository record.
    """
    body = {"name": name}
    effective_project_id = client._effective_project_id(project_id)
    if effective_project_id is not None:
        body["project_id"] = effective_project_id
    if dataset_name:
        body["dataset_name"] = dataset_name
    data = client._post("/api/vcs/repos", body)
    return Repo.from_dict(data)


def init_repo(
    client: RoteiroClient,
    name: str,
    project_id: Optional[int] = None,
    dataset_name: Optional[str] = None,
) -> Repo:
    """Backward-compatible alias for ``create_repo``."""
    return create_repo(
        client,
        name,
        project_id=project_id,
        dataset_name=dataset_name,
    )


def list_repos(
    client: RoteiroClient,
    project_id: Optional[int] = None,
) -> List[Repo]:
    """List managed repositories visible to the current tenant."""
    effective_project_id = client._effective_project_id(project_id)
    path = "/api/vcs/repos"
    if effective_project_id is not None:
        path = _with_query(path, {"project_id": effective_project_id})
    data = client._get(path)
    return [Repo.from_dict(repo) for repo in data]


def get_repo(client: RoteiroClient, repo_id: str) -> Repo:
    """Fetch a managed repository by ID."""
    data = client._get(f"/api/vcs/repos/{_encode_path_value(repo_id)}")
    return Repo.from_dict(data)


def delete_repo(client: RoteiroClient, repo_id: str) -> None:
    """Delete a managed repository by ID."""
    client._delete(f"/api/vcs/repos/{_encode_path_value(repo_id)}")


def commit(
    client: RoteiroClient,
    repo_id: str,
    input_path: str,
    message: str,
) -> Commit:
    """Create a new commit in a VCS repository.

    Converts the input dataset to GeoJSON (if needed) and stores it as a
    content-addressable blob in the repository.

    Args:
        client: An initialised RoteiroClient instance.
        repo_id: Managed repository ID.
        input_path: Path to the geospatial dataset to commit.
        message: Human-readable commit message.

    Returns:
        The created Commit object.
    """
    data = client._post(
        "/api/vcs/commit",
        {"repo_id": repo_id, "input": input_path, "message": message},
    )
    return Commit.from_dict(data)


def log(client: RoteiroClient, repo_id: str) -> List[Commit]:
    """Get the commit history for a repository.

    Returns commits ordered from most recent to oldest.

    Args:
        client: An initialised RoteiroClient instance.
        repo_id: Managed repository ID.

    Returns:
        A list of Commit objects.
    """
    data = client._get(_with_query("/api/vcs/log", {"repo_id": repo_id}))
    return [Commit.from_dict(c) for c in data]


def log_for_dataset(client: RoteiroClient, dataset_name: str) -> List[Commit]:
    """Get commit history for the managed repository linked to a dataset."""
    data = client._get(
        f"/api/vcs/log/dataset/{_encode_path_value(dataset_name)}"
    )
    return [Commit.from_dict(c) for c in data]


def diff(
    client: RoteiroClient,
    repo_id: str,
    commit_a: str,
    commit_b: str,
) -> DiffResult:
    """Compute the diff between two commits.

    Compares the geospatial snapshots at two commits and returns lists of
    added, removed, and modified features.

    Args:
        client: An initialised RoteiroClient instance.
        repo_id: Managed repository ID.
        commit_a: Commit ID or ref for the base (from) commit.
        commit_b: Commit ID or ref for the target (to) commit.

    Returns:
        A DiffResult with added, removed, modified features and stats.
    """
    data = client._get(
        _with_query(
            "/api/vcs/diff",
            {"repo_id": repo_id, "from": commit_a, "to": commit_b},
        )
    )
    return DiffResult.from_dict(data)


def checkout(
    client: RoteiroClient,
    repo_id: str,
    commit_id: str,
) -> None:
    """Check out a specific commit (restore a historical snapshot).

    This is a convenience method.  The VCS diff endpoint can be used to
    inspect what changed, and the commit's blob can be read directly for
    full restoration.

    Args:
        client: An initialised RoteiroClient instance.
        repo_id: Managed repository ID.
        commit_id: The commit ID to check out.

    Note:
        The current server-side implementation supports log and diff.
        Full checkout (restoring a snapshot) may require reading the
        commit blob directly.  This method retrieves the diff from the
        initial commit to the target as a lightweight proxy.
    """
    commits = log(client, repo_id)
    if not commits:
        from .client import RoteiroAPIError

        raise RoteiroAPIError("No commits in repository", status_code=404)

    # Find the oldest commit as the base reference.
    oldest = commits[-1]
    if oldest.id == commit_id:
        # Target is the initial commit; nothing to diff.
        return

    # Retrieve the diff to validate the commit exists.
    diff(client, repo_id, oldest.id, commit_id)
